rfd/0213-relay-a-lightweight-proxy.md

```diff
@@ -0,0 +1,66 @@
+
+This RFD will detail support for (agentful) SSH and Kubernetes, other protocols are out of scope - what _can_ be supported in the future varies based on the protocol and the class of client used to access it: everything should be supportable through a local `tsh proxy` (and/or Teleport Connect and/or Teleport VNet), direct client connections for Database access might be possible, App access seems harder to support (as it's heavily tied to the web UI, in its direct client connection mode).
+
+### Agent configuration and behavior
+
+Agents will be configured with a `proxy_server` address pointed at the usual Proxy public address (which is still going to be used for the initial joining process and for all Teleport API access), but they can optionally also include a `relay_server` pointing at the relay group address. If such an address is specified, agents will not open connections for the regular reverse tunnel. At least in the initial implementation it will only be valid to specify `relay_server` with `proxy_server` (and thus with a "v3" configuration file).
+
+Instead, agents configured to use a relay will connect to the relay group address to fetch the relay group discovery data (from any of the relays); this includes the relay group name, the agent connection count, and a list of the currently known relays for the group. They will then open reverse tunnel connections to different relays (up to the agent connection count) directly. This is different from the way that the regular reverse tunnel works, which requires agents to reach different proxies by opening new connections to the same public address to get randomly forwarded to different proxies behind a load balancer - connecting directly saves resources on the load balancer and the network, and it plays well with the intended usage scenario (agents and relays in the same subenvironment). Relays that are being gracefully terminated will report themselves as such to the connected agents (for both existing and new connections), and agents will not consider those relays for the purpose of filling the connection quota. Agents will include the relay group name and list of relay host IDs that they're connected to in their resources' heartbeats (the same way that proxy IDs are advertised in heartbeats today if proxy peering is enabled). Relays that are terminating will not be included in agents' heartbeats, reverse tunnel connections to terminating relays will be closed a few minutes after the last tunneled connection, and the relays themselves will wait a few minutes after the last tunneled connection before exiting. With appropriate grace times for disconnections as part of the graceful shutdown of the Teleport relay agents it should be straightforward to scale and upgrade a relay group without losing availability.
```

Comment:
option a: connect directly to the list of addresses
option b: connect via network load balancer via round robin

we will support option a and b.

Reply:
Default: load balancer, fall back direct to relay for tunnel

Reply:
I've punted on the direct connectivity and listed it as a possible future development since the PoC implementation ended up working out really well even if it's quite a bit simpler than even the existing reverse tunnel and the users we asked are fine with using a load balancer for now.

Reply:
Also, QUIC can migrate connections across different network paths, so if in the future we decide to enable QUIC for the tunnel we can even make connections try to move to a direct network path transparently, with the load balancer only steering the tunnel connections while they're being established.